### ISSUE TITLE
SHOULD or MUST intermediaries forward unknown capsule types without modification?

### ISSUE THREAD

**Issue by mirjak** [ietf-wg-masque/draft-ietf-masque-h3-datagram#167]
The first SHOULD should be a MUST because there is a clear exception specified but in all other cases this is a MUST.

The second SHOULD should really be a MUST. Or why is that a SHOULD?
If there is a reason for a SHOULD this needs to be explained but I think it would break interoperability.

**Comment by DavidSchinazi** [comment]
On #166, @mirjak said:
> I definitely missed this nuance during the wg call. What's the reason for a should?

**Comment by DavidSchinazi** [comment]
I think the main motivation was that the MUST was unenforceable, and that there will be intermediaries that explicitly want to block unknown extensions for security reasons.

**Comment by martinthomson** [comment]
Whatever we might prefer, isn't this a case where whatever the specification says, intermediaries will do what they want?

**Comment by DavidSchinazi** [comment]
I agree. That's why in the interest of making progress I'd rather move forward with the design team consensus instead of reopening this topic.

**Comment by martinthomson** [comment]
design team outputs don't have any special status here; perhaps the answer is not to say "SHOULD this" or "MUST that" but to instead articulate the intent behind the design and the potential consequences of not complying with that intent.

Perhaps the key reason for mandating forwarding is that the meaning of various protocol elements is only valid in the context of all of the messages that an endpoint says.  That is, you might send a bunch of datagram capsules that have meaning derived from other capsules.  Remove the other capsules and you change the meaning of those datagram capsules.  An intermediary that adds, modifies, or removes any capsules needs to understand the totality of the effect of their interventions.

This is why I remain skeptical about the value of the generic capsule protocol.  It only has value if it can be acted upon generically, but it is not obvious that it can be.  Put differently, the existence of a generic capsule protocol implies a number of constraints on the operation of extensions that are not obvious and not clearly articulated.

@gloinul might be asking about similar effects in #168.

**Comment by gloinul** [comment]
So lets start with explaining the interoperability issue that is at the core of this:

So the draft currently has several statements that interact with each other to cause this interoperability issue. 

Section 3.5 says: 

'Note that use of the Capsule Protocol is not required to use HTTP Datagrams. If an HTTP extension that uses HTTP Datagrams is only defined over transports that support QUIC DATAGRAM frames, it might not need a stream encoding.'

Section 3.5 also says: 

'An intermediary can reencode HTTP Datagrams as it forwards them. In other words, an intermediary MAY send a DATAGRAM capsule to forward an HTTP Datagram which was received in a QUIC DATAGRAM frame, and vice versa.'

Thus the following may occur:

Endpoint 1 sends a HTTP datagram as HTTP/3 Datagram over QUIC datagram. The HTTP intermediary re-encode this blind of its context to an Capsule Datagram and forward it to another HTTP intermediary that forwards it to an HTTP/3 endpoint using Datagram Capsule. As this HTTP datagram usage is one that does not require to support of Capsule Datagram this endpoint is not setup to handle this Datagram capsule and discards it. 

So this can either be fixed in various ways, not only what this issue text suggests:
   1. requiring the Datagram capsule support for all HTTP datagram handling nodes. In other words change the first quote.
   2. Require the HTTP intermediary to re-encode any Datagram capsules to HTTP/3 datagram if the next hop is know to support it. i.e. make additional requirements on how HTTP Intermediaries act when forwarding.
   3. Forbid blind forwarding of HTTP datagram in HTTP intermediaries.

I think we should settle this part before going back to see how the text Mirja suggested changing matter. I think 1 and 2 are both working solutions, and would like to avoid 3. 

I think there is also a question on the implication of what it means to support HTTP datagram when one is an HTTP/3 endpoint or intermediary. Who is expected to support the Datagram capsule if the next node is not capable?

**Comment by LPardue** [comment]
I actually think I prefer your option 3, caveated by saying "if you're an intermediary and you don't know the upgrade token, don't insert datagram capsules" . What concern do you have about option 3?

**Comment by gloinul** [comment]
Okay, I am skeptical against 3) on how it may make deployment of future extensions that require HTTP Datagram but doesn't need other than Intermediary forwarding. In that case deployment could be hampered if not all HTTP intermediaries are also HTTP/3 capable. It appears that using capsule protocol would be easier to enable over a legacy system. That might be a misconception from my part.

**Comment by DavidSchinazi** [comment]
@gloinul Thank you for clarifying, now I understand where the disconnect is. The draft currently says `Note that use of the Capsule Protocol is not required to use HTTP Datagrams.` and you interpreted that as `Note that the Capsule Protocol is OPTIONAL to implement for implementations of HTTP Datagrams.` but that's not what we intended, we meant `Note that it is possible for an HTTP extension to use HTTP Datagrams without using the Capsule Protocol.` I've written up #193 to remedy this

**Comment by gloinul** [comment]
@DavidSchinazi great that we managed to sort out what was the issue. As Lucas noted in the PR unless all HTTP Datagram implementation support and accept receiving Datagram capsule, even if not intended by the HTTP Extension, if unaware of HTTP extension transformation is allowed, it can fail. I think we need a bit more feedback on where people stand in this aspect.

**Comment by DavidSchinazi** [comment]
Thanks @gloinul. If the HTTP extension doesn't use capsules, then the implementation won't parse the data stream as capsules so there is no risk of accidentally receiving capsules. We added text to #193 to clarify the risks of not using the capsule protocol.

**Comment by LPardue** [comment]
I'd consider capsules to be similar to header fields (headers specifically, not trailers). Here's what HTTP says about them in https://httpwg.org/http-core/draft-ietf-httpbis-semantics-latest.html#rfc.section.5.1.p.6

>A proxy MUST forward unrecognized header fields unless the field name is listed in the [Connection](https://httpwg.org/http-core/draft-ietf-httpbis-semantics-latest.html#field.connection) header field ([Section 7.6.1](https://httpwg.org/http-core/draft-ietf-httpbis-semantics-latest.html#field.connection)) or the proxy is specifically configured to block, or otherwise transform, such fields. Other recipients SHOULD ignore unrecognized header and trailer fields. Adhering to these requirements allows HTTP's functionality to be extended without updating or removing deployed intermediaries.

Yeah, in actuality an intermediary can do lots of stuff, and in practice they do. But if a MUST is good enough for HTTP headers, why isn't it here?

**Comment by DavidSchinazi** [comment]
From memory it was either @martinthomson or @ekr who cared strongly about this not being a MUST. I'm fine either way as long as we don't drag this one needlessly.

**Comment by DavidSchinazi** [comment]
To summarize the discussion on this issue up to this point:
* we're debating whether to tell intermediaries that they MUST or SHOULD forward unknown capsules unmodified
* implementors of intermediaries will not react to either one of these words differently
* endpoints have no way of enforcing this MUST
* some folks felt very strongly against MUST
* the working group landed on this design [a month ago](https://mailarchive.ietf.org/arch/msg/masque/GXaxmCJoo122ymqyAqTUaUFix_k/) and no new information has surfaced since

Based on the points above, there is no reason to hold up document progress based on a single word that will not have any impact. Therefore I'm closing this issue with no action.
